### ISSUE TITLE
Check performance of chat with massive amount of messages

### ISSUE THREAD

**Issue by HenrikJannsen** [bisq-network/bisq2#948]

**Comment by namloan** [comment]
What's the success criteria? Is this about bursts of messages in a short period of time or about whether a long history of messages can degrade loading times? In a single chat window or adding up the messages from different chats?

**Comment by namloan** [comment]
> But probably best to set up a target scenario (e.g. 100 offers, 500 text messages) and then use a profiler to see where the bottlenecks are.

I'll do this

**Comment by HenrikJannsen** [comment]
I have seen that again a while back when I have many private messages in a trade that the UI got quite slow. So there is definitely an issue.

**Comment by namloan** [comment]
@HenrikJannsen what things have you noticed that go slower? With 100 offers and 1000+ messages there's a lag when clicking on 'offerbook' but once loaded it seems to work fine. I may need to give it a try on a VM with lower specs.

Early results from the profiling show that the heap keeps increasing every time the offerbook is loaded even if there are no new messages.

**Comment by namloan** [comment]
Heap going from 56 MB to 1 GB in ~5 mins:

https://github.com/bisq-network/bisq2/assets/138292162/a0b46be7-afd7-44e2-8922-d783b0d02893

**Comment by HenrikJannsen** [comment]
Seems there are some memory leaks. I remember to have seen in the item renderers some missing listener removals or other cleanups.
Which tool do you use?

**Comment by namloan** [comment]
I still need to analyze the profiling results, but looking at missing listener removals is a good hint.

> Which tool do you use?

[JDK Mission Control](https://www.oracle.com/java/technologies/jdk-mission-control.html)

**Comment by namloan** [comment]
From the automated analysis results:

The live set on the heap seems to increase with a speed of about 4.27 MiB per second during the recording.
An analysis of the reference tree found 2 leak candidates. The main candidate is bisq.desktop.main.content.bisq_easy.open_trades.BisqEasyOpenTradesController.view Referenced by this chain:
```
bisq.desktop.main.content.bisq_easy.open_trades.BisqEasyOpenTradesController$$Lambda$1721+0x0000000801701ab8.1781279162.arg$1
bisq.common.observable.collection.CollectionChangeHandler.listener
java.lang.Object[]
java.util.concurrent.CopyOnWriteArrayList.array
bisq.common.observable.collection.ObservableSet.observers
bisq.trade.bisq_easy.BisqEasyTradeStore.trades
bisq.trade.bisq_easy.BisqEasyTradeService.persistableStore
bisq.trade.TradeService.bisqEasyTradeService
bisq.desktop_app.DesktopApplicationService.tradeService
bisq.desktop_app.DesktopExecutable.applicationService
```

**Comment by namloan** [comment]
From another run (using ```-XX:FlightRecorderOptions=stackdepth=2048```, potentially more reliable as most of stack traces were being cut)

The live set on the heap seems to increase with a speed of about 604 KiB per second during the recording.
An analysis of the reference tree found 1 leak candidates. The main candidate is ```javafx.scene.control.skin.VirtualFlow.pile``` Referenced by this chain:
```
javafx.scene.control.ListView.top
bisq.desktop.main.content.components.ChatMessagesListView$View.listView
bisq.desktop.main.content.components.ChatMessagesListView$Controller.view
bisq.desktop.main.content.components.ChatMessagesListView$Controller$$Lambda$1243+0x00000008015be240.790136420.arg$1
org.fxmisc.easybind.EasyBind$$Lambda$465+0x0000000801390d58.1197210188.arg$1
javafx.beans.value.ChangeListener[]
com.sun.javafx.binding.ExpressionHelper$Generic.changeListeners
javafx.beans.property.ReadOnlyBooleanWrapper$ReadOnlyPropertyImpl.helper
javafx.stage.Window$5.readOnlyProperty
javafx.stage.Stage.focused
```

**Comment by namloan** [comment]
Could be related to the [bindChatMessages](https://github.com/bisq-network/bisq2/blob/cf28372feb0d18edbcf4477c1c9d0de0fbc34084/desktop/src/main/java/bisq/desktop/main/content/components/ChatMessagesListView.java#L455-L460) binding triggering too many updates - ```updateItem``` from the cell factory gets triggered followed by a bunch of CSS style updates, references stay alive in memory.

![bindchatmessages](https://github.com/bisq-network/bisq2/assets/138292162/16ffd545-aafa-479d-9b75-e1cd6622391f)

**Comment by HenrikJannsen** [comment]
ChatMessageListItem could benefit from caching, also for the values used in the compareTo and match methods.
I think best is to start to cleanup/refactor/optimize that part and then use the profiler to see if there is still some issues.

**Comment by namloan** [comment]
Started moving some object creations and styling updates out from ```updateItem``` and already noticed some improvements - slightly better loading times and the heap grows at a slower pace (couldn't get it past 400mb in a short session). I'll refactor to keep ```updateItem``` as light as possible. I could verify that it gets called a lot, orders of magnitude more than cells are created.

**Comment by namloan** [comment]
Performance dramatically improved, can easily deal with 8k+ messages in a single channel with barely any lag. I think there's not much value on further optimizations. Let me know what you think @HenrikJannsen 

For the record:

* Tried migrating to JavaFX 21. It seems more memory efficient (-50% thread allocation size), however in practice I noticed much more lag.
* I'd keep an eye on JavaFX 22 as it has further performance improvements like https://github.com/openjdk/jfx/pull/1076 . I tried the early access version but didn't notice any difference.
